### PR TITLE
Keyboard/kb04

### PR DESCRIPTION
**Pull request by filmstarr**
## Description

Added new keyboard KB04-01, made by DOIO. Extending https://github.com/qmk/qmk_firmware/pull/17739 to add RGB LEDs.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/doio/kb04/config.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+/* key matrix size */
+#define MATRIX_ROWS 1
+#define MATRIX_COLS 5
+
+/* key matrix pins */
+#define MATRIX_ROW_PINS { B3 }
+#define MATRIX_COL_PINS { B14, B13, B12, B0, A7 }
+
+#define BOOTMAGIC_LITE_ROW 0
+#define BOOTMAGIC_LITE_COLUMN 0
```
Comment (zvecr) [inline keyboards/doio/kb04/config.h:30]:
```suggestion
```

File: keyboards/doio/kb04/kb04.c [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ 
+ #include "kb04.h"
+
+#ifdef RGB_MATRIX_ENABLE
+led_config_t g_led_config = { {
+    {  0,      1,      2,      3,      4},
+}, {
+    {0  ,  0}, {75 ,  0}, {149,  0}, {224,  0}, {112,  64},
+}, {
+    1, 1, 1, 1, 1
```
Comment (zvecr) [inline keyboards/doio/kb04/kb04.c:25]: > RGB_MATRIX_LED_COUNT 8

If `RGB_MATRIX_LED_COUNT` is correct, these sections are 3 items too short.

File: keyboards/doio/kb04/kb04.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+ 
+ #pragma once
+
+#include "quantum.h"
+#include "print.h"
```
Comment (zvecr) [inline keyboards/doio/kb04/kb04.h:20]:
```suggestion
```

File: keyboards/doio/kb04/readme.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    `make doio/kb04:default`
+    or with the qmk cli:
+    `qmk compile -kb doio/kb04 -km default`
+
+Flashing example for this keyboard:
+
+    `make doio/kb04:default:flash`
+    or with the qmk cli:
+    `qmk flash -kb doio/kb04 -km default`
```
Comment (zvecr) [inline keyboards/doio/kb04/readme.md:21]:
```suggestion
    make doio/kb04:default

Flashing example for this keyboard:

    make doio/kb04:default:flash
```

File: keyboards/doio/kb04/rules.mk [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+NKRO_ENABLE = no           # Enable N-Key Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output
+
+# Additional thing to reduce compiled size
+LTO_ENABLE = yes
+SPACE_CADET_ENABLE = no
```
Comment (zvecr) [inline keyboards/doio/kb04/rules.mk:22]:
```suggestion
```

File: keyboards/doio/kb04/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    ),
+    [_LAY1] = LAYOUT(
+        KC_TRNS,     RGB_HUD,    RGB_TOG,    RGB_MOD,     KC_TRNS
+    ),
+    [_LAY2] = LAYOUT(
+        KC_TRNS,     KC_TRNS,    KC_TRNS,    KC_TRNS,     KC_TRNS
+    ),
+    [_LAY3] = LAYOUT(
+        KC_TRNS,     KC_TRNS,    KC_TRNS,    KC_TRNS,     KC_TRNS
+    ),
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/default/keymap.c:48]:
```suggestion
    )
```

File: keyboards/doio/kb04/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+};
+
+#ifdef ENCODER_MAP_ENABLE
+const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][2] = {
+    [_LAY0] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [_LAY1] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
+    [_LAY2] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
+    [_LAY3] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/default/keymap.c:56]:
```suggestion
    [_LAY1] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) }
```

File: keyboards/doio/kb04/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+ 
+ #include QMK_KEYBOARD_H
+
+enum layer_names {
+    _LAY0,
+    _LAY1,
+    _LAY2,
+    _LAY3
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/default/keymap.c:23]:
```suggestion
    _LAY1
```

**Review by zvecr — CHANGES_REQUESTED** [review]

**Comment by zvecr** [comment]
Compiles fine, as you can see with [39b1f4b](https://github.com/qmk/qmk_firmware/pull/19513/commits/39b1f4b7b62d20b46f6e557d0483252f9cf119b3).

**Comment by filmstarr** [comment]
Ahh, I understand! @zvecr, pulled your change and that works perfectly! 👌 Thanks for your patience.

File: keyboards/doio/kb04/readme.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
+# doio/kb04
+
+![doio/kb04](https://cdn.shopify.com/s/files/1/0463/5980/6101/products/4_f40396cc-ea72-47a5-887f-bbeeb4023b11_1100x.png?v=1654787686)
```
Comment (drashna) [inline keyboards/doio/kb04/readme.md:3]: Please rehost on Imgur.

Comment (filmstarr) [reply keyboards/doio/kb04/readme.md]: Imgur is not available to me (can't create an account). Have hosted it to Flickr instead.

**Review by drashna — APPROVED** [review]

File: keyboards/doio/kb04/config.h [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+ */
+
+#pragma once
+
+#include "config_common.h"
+
+/* key matrix size */
+#define MATRIX_ROWS 1
+#define MATRIX_COLS 5
+
```
Comment (fauxpark) [inline keyboards/doio/kb04/config.h:24]:
```suggestion
```

File: keyboards/doio/kb04/rules.mk [hunk 11]
```diff
@@ -0,0 +1,35 @@
+# MCU name
+MCU = STM32F103
+
+# Bootloader selection
+BOOTLOADER = stm32duino
+
```
Comment (fauxpark) [inline keyboards/doio/kb04/rules.mk:6]:
```suggestion
```

File: keyboards/doio/kb04/info.json [hunk 12]
```diff
@@ -0,0 +1,22 @@
+{
+    "keyboard_name": "KB04-01",
+    "manufacturer": "DOIO",
+    "url": "https://www.keebmonkey.com/products/megalodon-macro-pad-with-a-knob",
+    "maintainer": "filmstarr",
+    "usb": {
+        "vid": "0xD010",
+        "pid": "0x0401",
+        "device_version": "0.0.1"
+    },
```
Comment (fauxpark) [inline keyboards/doio/kb04/info.json:10]:
```suggestion
    },
    "processor": "STM32F103",
    "bootloader": "stm32duino",
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/doio/kb04/info.json [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
```
Comment (fauxpark) [inline keyboards/doio/kb04/info.json:10]:
```suggestion
    },
    "matrix_pins": {
        "cols": ["B14", "B13", "B12", "B0", "A7"],
        "rows": ["B3"]
    },
    "diode_direction": "COL2ROW",
    "encoder": {
        "rotary": [
            {"pin_a": "B5", "pin_b": "B6"},
        ]
    },
```

File: keyboards/doio/kb04/config.h [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCE 5
+
+/* Encoder pins */
+#define ENCODERS_PAD_A { B5 }
+#define ENCODERS_PAD_B { B6 }
+
```
Comment (fauxpark) [inline keyboards/doio/kb04/config.h:34]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by zqad** [comment]
Hi,
For what it's worth, I tested this branch with my KB04 and it works just perfectly with every keymap and LED setup I threw at it. Thanks @filmstarr!

**Comment by zqad** [comment]
Hi,
Anything that I could help test or so, before this PR gets auto-closed? Would be nice to get this merged :)

File: keyboards/doio/kb04/keymaps/via/keymap.c [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    ),
+    [_LAY1] = LAYOUT(
+        KC_TRNS,     RGB_HUD,    RGB_TOG,    RGB_MOD,     KC_TRNS
+    ),
+    [_LAY2] = LAYOUT(
+        KC_TRNS,     KC_TRNS,    KC_TRNS,    KC_TRNS,     KC_TRNS
+    ),
+    [_LAY3] = LAYOUT(
+        KC_TRNS,     KC_TRNS,    KC_TRNS,    KC_TRNS,     KC_TRNS
+    ),
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/via/keymap.c:48]:
```suggestion
    )
```

File: keyboards/doio/kb04/keymaps/via/keymap.c [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+};
+
+#ifdef ENCODER_MAP_ENABLE
+const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][2] = {
+    [_LAY0] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [_LAY1] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
+    [_LAY2] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
+    [_LAY3] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/via/keymap.c:56]:
```suggestion
const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][NUM_DIRECTIONS] = {
    [_LAY0] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
    [_LAY1] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) }
```

File: keyboards/doio/kb04/rules.mk
Comment (zvecr) [inline keyboards/doio/kb04/rules.mk:1]: Most of this should be in info.json

File: keyboards/doio/kb04/keymaps/via/keymap.c [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
+enum layer_names {
+    _LAY0,
+    _LAY1,
+    _LAY2,
+    _LAY3
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/via/keymap.c:23]:
```suggestion
    _LAY1
```

File: keyboards/doio/kb04/kb04.c
Comment (zvecr) [inline keyboards/doio/kb04/kb04.c:1]: This can be configured in info.json, and then the file can be removed.

File: keyboards/doio/kb04/config.h [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#ifdef RGB_MATRIX_ENABLE
+    /* RGB Matrix config */
+    #define RGB_DI_PIN A10
```
Comment (zvecr) [inline keyboards/doio/kb04/config.h:21]: `RGB_DI_PIN` is now invalid, and would be `ws2812.pin` in info.json (or  `WS2812_DI_PIN`)

**Review by zvecr — CHANGES_REQUESTED** [review]

**Comment by vinorodrigues** [comment]
#22754

**Pull request closed without merging**